engine/sim/sc_expressions.cpp

```diff
@@ -16,6 +16,145 @@ namespace
 {  // ANONYMOUS ====================================================
 
 constexpr bool EXPRESSION_DEBUG = false;
+
+struct lexer_t
+{
+  struct token_t {
+    token_e type = TOK_UNKNOWN;
+    util::string_view str;
+  };
+
+  util::string_view input;
+  size_t current_len = 0;
+
+  explicit lexer_t( util::string_view s )
+    : input( s )
+  {}
+
+  char peek() const
+  {
+    return current_len < input.size() ? input[ current_len ] : '\0';
+  }
+
+  bool match( char c )
+  {
+    if ( starts_with( input.substr( current_len ), c ) ) {
+      current_len++;
+      return true;
+    }
+    return false;
+  }
+
+  bool match( util::string_view str, int consume_mod = 0 )
+  {
+    if ( util::str_prefix_ci( input.substr( current_len ), str ) ) {
+      current_len += ( str.size() + consume_mod );
```

Comment:
Maybe add an assert here that the resulting calculation is still in bounds of the input.
I understand that for the current match calls with consume_mod = -1 this should always be fine.